xdsl/dialects/fsm.py

```diff
@@ -0,0 +1,377 @@
+                )
+        if self.res_attrs is not None and self.res_names is not None:
+            if len(self.res_attrs) != len(self.res_names):
+                raise VerifyException(
+                    "The number of res_attrs and res_names should be the same"
+                )
+
+
+@irdl_op_definition
+class Output(IRDLOperation):
+    """Represents the outputs of a machine under a specific state. The
+    types of `$operands` should be consistent with the output types of the state
+    machine"""
+
+    name = "fsm.output"
+
+    operand = var_operand_def(AnyAttr())
+
+    traits = frozenset([IsTerminator()])
+
+    def __init__(
+        self,
+        operand: Sequence[SSAValue | Operation],
+    ):
+        super().__init__(
+            operands=[operand],
+        )
+
+    def verify_(self):
+        parent = self.parent_op()
+        if (
+            isinstance(parent, State)
+            and parent.transitions == self.parent_region()
+            and len(self.operands) > 0
+        ):
+            raise VerifyException("Transition regions should not output any value")
+        while parent is not None:
+            if isinstance(parent, Machine):
+                if not (
+                    [operand.type for operand in self.operands]
+                    == [type(result) for result in parent.function_type.outputs]
+                    and len(self.operands) == len(parent.function_type.outputs)
+                ):
+                    raise VerifyException(
+                        "Output types must be consistent with the machine's"
+                    )
+            parent = parent.parent_op()
+
+
+@irdl_op_definition
+class State(IRDLOperation):
+    """Represents a state of a state machine. This op includes an
+    `$output` region with an `fsm.output` as terminator to define the machine
+    outputs under this state. This op also includes a `transitions` region that
+    contains all the transitions of this state"""
+
+    name = "fsm.state"
+
+    # includes output and transitions region
+
+    output = region_def()
+
+    transitions = region_def()
+
+    # attributes
+
+    sym_name = attr_def(StringAttr)
+
+    traits = frozenset([NoTerminator(), SymbolOpInterface()])
+
+    def __init__(
+        self,
+        sym_name: str,
+        output: Region | type[Region.DEFAULT] = Region.DEFAULT,
+        transitions: Region | type[Region.DEFAULT] = Region.DEFAULT,
+    ):
+        attributes: dict[str, Attribute] = {}
+        attributes["sym_name"] = StringAttr(sym_name)
+        if not isinstance(output, Region):
+            output = Region(Block())
+        if not isinstance(transitions, Region):
+            transitions = Region(Block())
+        super().__init__(
+            attributes=attributes,
+            regions=[output, transitions],
+        )
+
+    def verify_(self):
+        parent = self.parent_op()
+
+        while parent is not None:
```

Comment:
Can there be MachineOps nested inside other MachineOps? If not, you would not need to check all parents, as you are doing here :)

Reply:
I had a look at the rationale document for FSM and they say it was inspired by UML state machines.
AFAIR, UML has the notion of nesting in state machines, so maybe we need to verify this with a basic nesting example in CIRCT.

Regarding the second point, `StateOp` needs to have a `MachineOp` as a parent since it specifies the `HasParent` trait. The doc mentions only the trait without mention to the operation that needs to fulfill this relationship, but looking at the code we have:

```
def StateOp : FSMOp<"state", [HasParent<"MachineOp">, Symbol, NoTerminator]> {
```

Reply:
Thanks a lot! super useful suggestions